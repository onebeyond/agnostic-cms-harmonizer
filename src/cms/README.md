## CLI
Install [Contentful CLI](https://www.contentful.com/developers/docs/tutorials/cli/import-and-export/#requirements) or alternatively use `npx contentful`

## Environment Variables
 
 1. You have to create a space in contentful to run the tests.
 2. Follow the instructions [here](https://www.contentful.com/developers/docs/references/authentication/#the-content-delivery-and-preview-api) to create an API key.
 3. Setup your `.env` file as follows

```text
CONTENTFUL_ACCESS_TOKEN=<your_access_token>
CONTENTFUL_ENVIRONMENT=<your_environment>
CONTENTFUL_SPACE=<your_space>
CONTENTFUL_ENTRY=<your_entry>
```

## Import dummy content to your space
To run the e2e tests you need to be authenticated in your contentful account and import the dummy content into the space you created before. The content you need to import is located at __tests__/exports/contentful/agnostic-cms-harmonizer_space.json and should not be updated.

To log in to your contentful account run the following command and follow the instructions in the terminal:

# If you have the contenful-cli installed globally
contenful login

# If you don't have the contenful-cli installed globally
npx contentful-cli login
To import the dummy content run the following command:

# If you have the contenful-cli installed globally
contentful space import --content-file __tests__/exports/contentful/agnostic-cms-harmonizer_space.json --space-id <your-contentful-space-id> --environment-id <your-contentful-environment-id>

# If you don't have the contenful-cli installed globally
npx contentful-cli space import --content-file __tests__/exports/contentful/agnostic-cms-harmonizer_space.json --space-id <your-contentful-space-id> --environment-id <your-contentful-environment-id>
There is also a located at ./__tests__/exports/contentful/config.json in case you need to export your space's content for any reason. You just need to run the following command:

# If you have the contenful-cli installed globally
contentful space export --config ./__tests__/exports/contentful/config.json --space-id <your-contentful-space-id> --environment-id <your-contentful-environment-id>

# If you don't have the contenful-cli installed globally
npx contentful-cli space export --config ./__tests__/exports/contentful/config.json --space-id <your-contentful-space-id> --environment-id <your-contentful-environment-id>
