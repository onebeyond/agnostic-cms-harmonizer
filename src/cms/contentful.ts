import {
  createClient,
  type CreateClientParams,
  type ContentfulClientApi,
  type Entry,
  type EntrySkeletonType,
  type EntryQueries,
  type FieldsType,
  type ChainModifiers,
  EntriesQueries,
} from 'contentful';

import { HarmonizedOutput } from '../@types';
import {
  AbstractAgnosticCMSHarmonizerClient,
  AbstractGetEntriesParams,
  AbstractGetEntryParams,
} from '../index.abstract';

/**
 * Type of entries that can be retrieved
 * @internal
 */
const ContentfulResourceType = {
  ASSET: 'Asset',
  ENTRY: 'Entry',
  LINK: 'Link',
} as const;

export type ContentfulEntry<T> = Entry<EntrySkeletonType<T & FieldsType>, undefined, string>;

export type ContentfulEntrySkeleton<T> = EntrySkeletonType<T & FieldsType>;

export type ContentfulGetEntryParams = AbstractGetEntryParams &
  EntryQueries<ChainModifiers> & {
    locale?: string;
    nestedLevels?: 0 | 2 | 1 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10;
  };

export type ContentfulGetEntriesParams<T> = AbstractGetEntriesParams &
  EntriesQueries<ContentfulEntrySkeleton<T>, ChainModifiers> & {
    locale?: string;
    nestedLevels?: 0 | 2 | 1 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10;
  };

/**
 * Contentful provider.
 */
export class HarmonizerContentfulClient extends AbstractAgnosticCMSHarmonizerClient {
  /**
   * @param clientParams {@link https://contentful.github.io/contentful.js/contentful/10.6.21/interfaces/CreateClientParams.html}
   */
  constructor(clientParams: CreateClientParams) {
    super(clientParams);
    this.clientInstance = Object.create(null);
  }

  public async initialize(): Promise<void> {
    this.clientInstance = await this.agnosticCmsInitialize(async () =>
      createClient(this.clientParams),
    );
  }

  /**
   * {@link AbstractAgnosticCMSHarmonizerClient#getEntry}
   * @param {ContentfulGetEntryParams} params
   * @remarks
   * By providing the `entryId` parameter, you can fetch the data for a specific _Contentful entry_.
   * Additionally, you can optionally specify the `locale` and `nestedLevels` parameters.
   * The `nestedLevels` parameter determines the depth of _reference resolution_ in the entry and
   * has a default value of __10__. You can also specify the expected data type by using a _type argument_.
   * @returns {Promise<HarmonizedOutput<T>>}
   * @example
   * type MyEntry = {
   *  title: string;
   *  description: string;
   * }
   *
   * const client = await (new HarmonizerContentfulClient({...})).initialize();
   * const entry = await client.getEntry<MyEntry>({ entryId: '123', locale: 'en-US' });
   * console.log(entry); // { data: { title: 'My title', description: 'My description' } }
   */
  public async getEntry<T = Record<string, unknown>>({
    entryId,
    locale,
    nestedLevels = 10,
  }: ContentfulGetEntryParams): Promise<HarmonizedOutput<T>> {
    const query = {
      locale,
      include: nestedLevels,
    };

    return await this.getEntryHarmonized<T>(
      <() => Promise<T>>this.getEntryHandler?.bind(this, entryId, query),
      <(data: T) => HarmonizedOutput<T>>this.parserHandler.bind(this),
    );
  }

  public async getEntries<T = Record<string, unknown>>({
    collectionId,
    locale,
    nestedLevels = 10,
    ...config
  }: ContentfulGetEntriesParams<T>): Promise<HarmonizedOutput<T[]>> {
    const query = {
      ...config,
      locale,
      content_type: collectionId,
      include: nestedLevels,
    };
    return await this.getEntriesHarmonized<T>(
      <() => Promise<T[]>>this.getEntriesHandler?.bind(this, query),
      <(data: T[]) => HarmonizedOutput<T[]>>this.parserHandler.bind(this),
    );
  }

  protected clientInstance: ContentfulClientApi<undefined>;

  protected getClientInstance(): ContentfulClientApi<undefined> {
    return this.clientInstance;
  }

  private async getEntryHandler<T = Record<string, unknown>>(
    entryId: string,
    query?: EntryQueries<undefined>,
  ): Promise<ContentfulEntry<T>> {
    return this.getClientInstance().getEntry<ContentfulEntrySkeleton<T>>(entryId, query);
  }

  private async getEntriesHandler<T = Record<string, unknown>>(
    query?: EntriesQueries<ContentfulEntrySkeleton<T>, undefined>,
  ): Promise<ContentfulEntry<T>[]> {
    const entries = await this.getClientInstance().getEntries<ContentfulEntrySkeleton<T>>(query);
    return entries.items;
  }

  private parserHandler<T = Record<string, unknown>>(
    payload: ContentfulEntry<T> | ContentfulEntry<T>[],
  ): HarmonizedOutput<string | ContentfulEntry<T> | ContentfulEntry<T>[] | null> {
    if (Array.isArray(payload)) {
      return {
        data: payload.reduce((items, item) => {
          return item ? items.concat(this.mapper<T>(item)) : items;
        }, Array(0)),
      };
    }

    return { data: this.mapper<T>(payload) };
  }

  private mapper<T = Record<string, unknown>>(
    item: ContentfulEntry<T>,
  ): string | ContentfulEntry<T> | null {
    switch (item?.sys?.type + '') {
      case ContentfulResourceType.ENTRY:
        return Object.entries(item.fields).reduce((acc, [key, value]) => {
          if (Array.isArray(value)) {
            return {
              ...acc,
              [key]: value.map((valueItem) => this.mapper<T>(valueItem || Object.create(null))),
            };
          }
          return {
            ...acc,
            [key]: this.mapper<T>(value || Object.create(null)),
          };
        }, Object.create(null));

      case ContentfulResourceType.ASSET:
        return Object(item.fields?.file)?.url ? `https:${Object(item.fields.file).url}` : null;

      case ContentfulResourceType.LINK:
        return Object.assign(Object.create(null), { id: item.sys.id });

      default:
        return item;
    }
  }
}
