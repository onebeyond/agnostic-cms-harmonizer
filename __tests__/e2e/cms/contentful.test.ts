import { config as configDotenv } from 'dotenv';
configDotenv();

import { Contentful } from '../../../src/cms/contentful';

describe('contentful', () => {
  let contentful: Contentful;

  beforeAll(async () => {
    contentful = new Contentful({
      accessToken: process.env.CONTENTFUL_ACCESS_TOKEN as string,
      environment: process.env.CONTENTFUL_ENVIRONMENT as string,
      space: process.env.CONTENTFUL_SPACE as string,
    });
    await contentful.initialize();
  });

  describe('should not fail', () => {
    it('retrieve content in default locale (en-US)', async () => {
      const harmonizedData = await contentful.getEntry({
        entryId: process.env.CONTENTFUL_ENTRY as string,
      });
      expect(harmonizedData).toMatchSnapshot({
        data: {
          singleMediaInputField: expect.stringMatching(
            /^https:\/\/images\.ctfassets\.net\/.+\/banner\.jpg$/,
          ),
        },
      });
    });

    it('retrieve only root level of content in default locale (en-US)', async () => {
      const harmonizedData = await contentful.getEntry({
        entryId: process.env.CONTENTFUL_ENTRY as string,
        nestedLevels: 0,
      });
      expect(harmonizedData).toMatchSnapshot();
    });

    it('retrieve content in specific locale `es`', async () => {
      const harmonizedData = await contentful.getEntry({
        entryId: process.env.CONTENTFUL_ENTRY as string,
        locale: 'es',
      });
      expect(harmonizedData).toMatchSnapshot({
        data: {
          singleMediaInputField: expect.stringMatching(
            /^https:\/\/images\.ctfassets\.net\/.+\/banner\.jpg$/,
          ),
        },
      });
    });
  });
});
