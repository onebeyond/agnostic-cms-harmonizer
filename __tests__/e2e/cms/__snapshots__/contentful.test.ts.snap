// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contentful should not fail retrieve array of available entries in default locale (en-US) 1`] = `
{
  "data": [
    {
      "booleanInputFieldDefaultTrue": true,
      "jsonInputField": {
        "objectKey": "objectValue",
      },
      "questions": [
        {
          "configuration": {
            "identifier": "First name question configuration",
            "required": true,
          },
          "identifier": "First name question",
          "label": "First name",
          "placeholder": "Please enter your name",
        },
      ],
      "simpleShortInputTextAreaOne": "Amidst the chaos of dreams, a whisper of serenity finds its way.",
      "singleMediaInputField": "https://images.ctfassets.net/tfwqn4ih0op2/36j3phOyWhdSaqhRKeLauu/131db2ba1136b43a16e1a0f4f046c6a0/banner.jpg",
    },
  ],
}
`;

exports[`contentful should not fail retrieve array of available entries in given locale (es) 1`] = `
{
  "data": [
    {
      "booleanInputFieldDefaultTrue": true,
      "jsonInputField": {
        "objectKey": "objectValue",
      },
      "questions": [
        {
          "configuration": {
            "identifier": "First name question configuration",
            "required": true,
          },
          "identifier": "First name question",
          "label": "Nombre",
          "placeholder": "Por favor, introduzca su nombre",
        },
      ],
      "simpleShortInputTextAreaOne": "Amidst the chaos of dreams, a whisper of serenity finds its way.",
      "singleMediaInputField": "https://images.ctfassets.net/tfwqn4ih0op2/36j3phOyWhdSaqhRKeLauu/8e57bd83434abf21220c1fb5d75efdda/banner.jpg",
    },
  ],
}
`;

exports[`contentful should not fail retrieve content in default locale (en-US) 1`] = `
{
  "data": {
    "booleanInputFieldDefaultTrue": true,
    "jsonInputField": {
      "objectKey": "objectValue",
    },
    "questions": [
      {
        "configuration": {
          "identifier": "First name question configuration",
          "required": true,
        },
        "identifier": "First name question",
        "label": "First name",
        "placeholder": "Please enter your name",
      },
    ],
    "simpleShortInputTextAreaOne": "Amidst the chaos of dreams, a whisper of serenity finds its way.",
    "singleMediaInputField": StringMatching /\\^https:\\\\/\\\\/images\\\\\\.ctfassets\\\\\\.net\\\\/\\.\\+\\\\/banner\\\\\\.jpg\\$/,
  },
}
`;

exports[`contentful should not fail retrieve content in specific locale \`es\` 1`] = `
{
  "data": {
    "booleanInputFieldDefaultTrue": true,
    "jsonInputField": {
      "objectKey": "objectValue",
    },
    "questions": [
      {
        "configuration": {
          "identifier": "First name question configuration",
          "required": true,
        },
        "identifier": "First name question",
        "label": "Nombre",
        "placeholder": "Por favor, introduzca su nombre",
      },
    ],
    "simpleShortInputTextAreaOne": "Amidst the chaos of dreams, a whisper of serenity finds its way.",
    "singleMediaInputField": StringMatching /\\^https:\\\\/\\\\/images\\\\\\.ctfassets\\\\\\.net\\\\/\\.\\+\\\\/banner\\\\\\.jpg\\$/,
  },
}
`;

exports[`contentful should not fail retrieve only root level of content in default locale (en-US) 1`] = `
{
  "data": {
    "booleanInputFieldDefaultTrue": true,
    "jsonInputField": {
      "objectKey": "objectValue",
    },
    "questions": [
      {
        "id": "5vX5fGs6Px3Khf1JO9blyW",
      },
    ],
    "simpleShortInputTextAreaOne": "Amidst the chaos of dreams, a whisper of serenity finds its way.",
    "singleMediaInputField": {
      "id": "36j3phOyWhdSaqhRKeLauu",
    },
  },
}
`;
